ggu/main.c

```diff
@@ -16,11 +16,380 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <stdbool.h>
+
+#include <rte_ip.h>
+#include <rte_udp.h>
+#include <rte_log.h>
+#include <rte_lcore.h>
+#include <rte_malloc.h>
+#include <rte_ethdev.h>
+#include <rte_atomic.h>
+
+#include "gatekeeper_gk.h"
 #include "gatekeeper_ggu.h"
+#include "gatekeeper_net.h"
+#include "gatekeeper_main.h"
+#include "gatekeeper_config.h"
+
+#define GGU_PD_VER1 (1)
+
+static rte_atomic16_t num_ggu_conf_alloc = RTE_ATOMIC16_INIT(0);
+
+static void
+process_single_policy(struct ggu_policy *policy, struct ggu_config *ggu_conf)
+{
+	/*
+	 * Identify which GK block is responsible for the
+	 * pair <Src, Dst> in the decision.
+	 */
+	int block_id = get_responsible_block(&policy->flow, ggu_conf->gk);
+	struct gk_cmd_entry entry;
+
+	if (block_id < 0)
+		return;
+	/*
+	 * Obtain mailbox of that GK block,
+	 * and send the policy decision to the GK block.
+	 */
+	entry.op = GGU_POLICY_ADD;
+	entry.u.ggu.state = policy->state;
+	rte_memcpy(&entry.u.ggu.flow, &policy->flow, sizeof(policy->flow));
```

Comment:
It's always safer to put the destination field (i.e. `entry.u.ggu.flow`) in the `sizeof()`. The reason being this guarantees that one can't ever go over buffer.